### PR TITLE
Allow bump_rpm to update specfile

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

File: bump_rpm.sh
Comment (ekohl) [inline bump_rpm.sh:86]:
```suggestion
		if [[ $TEMPLATE == 'scl' || $TEMPLATE == 'nonscl' || -z $TEMPLATE ]] ; then
```

File: bump_rpm.sh
Comment (ekohl) [inline bump_rpm.sh:96]: This unconditionally always rewrites the entire spec, but that wipes out local changes. I think this should be hidden behind an env var.

A safer way could be that it only rewrites if the template is `scl`, `nonscl` or missing, otherwise falls back to the safer replacement.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: bump_rpm.sh
Comment (ehelms) [reply bump_rpm.sh:96]: What scenario do you forsee where you don't want the spec file getting updated? or that you have local changes where you want to bump the RPM?

Comment (ekohl) [reply bump_rpm.sh:96]: It is mostly about the spec files where the template is insufficient (https://github.com/theforeman/foreman-packaging/pull/8166 comes to mind). I don't think we'll ever capture everything and we do have a bot now that submits PRs using the script. So I'd stay on the safe side.

File: packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec
Comment (ekohl) [inline packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec:8]: We could drop the group from the template. Fedora has done that already and it doesn't add anything.

File: packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec
Comment (ekohl) [inline packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec:22]: This is an example of where the template is insufficient. I've opened https://github.com/moiristo/deep_cloneable/pull/142 but for now:
```suggestion
that includes user specified associations.
```

File: packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec
Comment (ekohl) [inline packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec:67]: Should we exclude this file?

File: packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec
Comment (ekohl) [inline packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec:69]: I wonder why this shows up. We have `%{gem_spec} already so why is it duplicated?

**Review by ekohl — COMMENTED** [review]
Did you intend to include the deep_clonable update in the same commit?

File: packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec
Comment (ehelms) [reply packages/foreman/rubygem-deep_cloneable/rubygem-deep_cloneable.spec:69]: So far, I have not been able to track down why the logic is grabbing this as a `doc_files`

**Comment by ehelms** [comment]
I added a drop of Group and ignore on Gemfile.lock. This also includes a commit that updates a few of the pre/build/setup sections to match Fedora template as I think our steps were a bit out dated.

File: bump_rpm.sh
Comment (ekohl) [reply bump_rpm.sh:96]: > What is the end goal there?

I'd say that it allows a conservative update by default but you can force to fully regenerate specs (on the next update) by increasing the template's version.

> Sorry, I know you said this earlier but I realize I do not know what you mean by this.

Essentially:
```sh
if [[ $TEMPLATE == 'scl' || $TEMPLATE == 'nonscl' || -z $TEMPLATE ]] ; then
  regenerate_spec_file
elif [[ -n $TEMPLATE ]] ; then
  update_spec_file
fi
```

Comment (ehelms) [reply bump_rpm.sh:96]: What is the difference between regenerate and update in this context? Are you implying that we would have something like:

```
gem2rpm/default-latest.spec.erb
gem2rpm/default-2.spec.erb
gem2rpm/default-1.spec.erb
```

And if a spec file had `# template: default 1` the regenerate action would use this template and the update action would use the latest template?

I am getting a bit lost in the nuance here and maybe that is a sign that we are overengineering things a bit. Whenever we update the version of a gem we know that there are things that may change:

 * requirements (build or runtime)
 * new files that may need to be ignored

Those things are captured by doing a specfile update.Then we have some special case scenarios where some gems may have extracurricular additions to their spec file that break the norm. My argument is this is already a problem that we run into anyway and we should incrementally start adapting to this by first accepting that we should also run a specfile update anytime we do a version bump. The script and bot do the hard work of detecting updates and opening PRs, maintainers still should review and look for anomalies and then take over the PR with updates as needed. That feels sane to me while we engineer a way to preserve the extracurricular's. That's how I looked at how we might incrementally and iteratively automate this more and more to reduce packaging burden.

Comment (ekohl) [reply bump_rpm.sh:96]: My biggest worry is that we can't rely on gem2rpm to generate a 100% correct spec, so we must make some (small) modifications. If every time the bot tries to update it those get reverted we actually get a lot of additional work. This is why I initially wrote the `update-requirements` script the way I did: allow it to replace specific parts. If we decide that files should also be replaced we can also add similar markers to `%files` to replace those bits automatically.

Then we also have cases where we do want to regenerate the entire spec file. Getting rid of the SCL macros is an excellent example: it doesn't happen often, but when it does the automation helps us a lot.

That's why I'm leaning to a cautious approach: essentially opt-in full replacement. Right now we have a case where we know it's probably the right thing to do ((non)scl to default). My versioning thoughts were about how we can do a similar thing in the future.

So short term I'd suggest:
```bash
                if [[ $TEMPLATE == 'scl' ]] || [[ $TEMPLATE == 'nonscl' ]] || [[ -z $TEMPLATE ]]; then
                        CHANGELOG=$(mktemp)
                        trap "rm -f $CHANGELOG" EXIT
                        sed -e '1,/%changelog/ d' $SPEC_FILE > $CHANGELOG
                        gem2rpm -t $ROOT/gem2rpm/default.spec.erb -o $SPEC_FILE *.gem
                        cat $CHANGELOG >> $SPEC_FILE
                        git add $SPEC_FILE
                elif [[ -n $TEMPLATE ]] ; then
                        echo "* Updating requirements"
                        gem2rpm -t $ROOT/gem2rpm/$TEMPLATE.spec.erb *.gem | $SCRIPT_DIR/update-requirements specfile - $SPEC_FILE
                        git add $SPEC_FILE
                fi
```

Then later, when we do want to force regeneration of the default template, we can think about another trigger to force regeneration.

To improve on regeneration we can change the files section to:
```erb
%files
# start specfile generated files main
%dir %{gem_instdir}
....
%{gem_spec}
# end specfile generated files main
```

And the same with docs:
```erb
<% if doc_subpackage -%>
%files doc
# start specfile generated files doc
%doc %{gem_docdir}
<%= doc_files.to_rpm %>
# end specfile generated files doc
<% end # if doc_subpackage -%>
```

This way `update-requirements` can replace files too.

Comment (ehelms) [reply bump_rpm.sh:96]: A thought I had been noodling on -- https://github.com/theforeman/foreman-packaging/pull/8169

Comment (ekohl) [reply bump_rpm.sh:96]: How about we start with https://github.com/theforeman/foreman-packaging/pull/8206 and then see where it goes from there?

**Comment by ekohl** [comment]
Could you rebase this on top of `rpm/develop`? I do like the spec file cleanups which I didn't include in my PR.

**Review by ekohl — APPROVED** [review]
I didn't check if packages build with it, but that should be easy enough to figure out.

**Comment by ehelms** [comment]
[test rpm]

**Pull request merged**
